src/mame/machine/eispc_kb.cpp

```diff
@@ -0,0 +1,446 @@
+	PORT_BIT( 0x2000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_S)          PORT_CHAR('s') PORT_CHAR('S')                  PCM(key, 5) // 31
+	PORT_BIT( 0x4000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_B)          PORT_CHAR('b') PORT_CHAR('B')                  PCM(key, 5) // 48
+	PORT_BIT( 0x8000, IP_ACTIVE_LOW, IPT_KEYBOARD )                     PORT_CODE(KEYCODE_N)          PORT_CHAR('n') PORT_CHAR('N')                  PCM(key, 5) // 49
+INPUT_PORTS_END
+
+
+//-------------------------------------------------
+//  ROM( eispc_kb )
+//-------------------------------------------------
+
+ROM_START( eispc_kb )
+	ROM_REGION( 0x800, M6801_TAG, 0 )
+	ROM_LOAD( "sgtpepper-1.2.bin", 0x000, 0x800, CRC(7107b841) SHA1(a939dd50622575c31fea9c7adb7a7db5403a7aca) )
+ROM_END
+
+} // anonymous namespace
+
+//**************************************************************************
+//  DEVICE DEFINITIONS
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(EISPC_KB, eispc_keyboard_device, "eispc_kb", "Ericsson PC keyboard")
+
+//**************************************************************************
+//  LIVE DEVICE
+//**************************************************************************
+
+//-------------------------------------------------
+//  eispc_keyboard_device - constructor
+//-------------------------------------------------
+
+eispc_keyboard_device::eispc_keyboard_device(
+		machine_config const &mconfig,
+		char const *tag,
+		device_t *owner,
+		uint32_t clock)
+	: device_t(mconfig, EISPC_KB, tag, owner, clock)
+	, m_mcu(*this, M6801_TAG)
+	, m_rows(*this, "P1%u", 0)
+	, m_txd_cb(*this)
+	, m_rxd_high(true)
+	, m_txd_high(true)
+	, m_hold(true)
+	, m_col_select(0)
+{
+}
+
+INPUT_CHANGED_MEMBER( eispc_keyboard_device::key )
+{
+		if (oldval && !newval)
+	{
+		LOGUI("Key Pressed - name: %s field: %04x param: %04x oldval: %04x newval: %04x\n", field.name(), field.defvalue(), param, oldval, newval);
+		int idx = *((int *)(&param));
+		if (idx >= sizeof(keys))
+			logerror("Out of bounds access in keys array\n");
+		else
+			keys[idx] |= (uint16_t) field.defvalue();
+	}
+	else if (newval && !oldval)
+	{
+		LOGUI("Key Released - name: %s field: %04x param: %04x oldval: %04x newval: %04x\n", field.name(), field.defvalue(), param, oldval, newval);
+		int idx = *((int *)(&param));
+		if (idx >= sizeof(keys))
+			logerror("Out of bounds access in keys array\n");
+		else
+			keys[idx] &= ~(uint16_t)field.defvalue();
+	}
+	for (int i = 0; i < 6; i++) LOGUI("%04x ", keys[i]); LOGUI("\n");
+}
+
+WRITE_LINE_MEMBER(eispc_keyboard_device::rxd_w)
+{
+	LOGBITS("KBD bit presented: %d\n", state);
+	m_rxd_high = CLEAR_LINE != state;
+}
+
+WRITE_LINE_MEMBER(eispc_keyboard_device::hold_w)
+{
+	m_hold = CLEAR_LINE == state;
+}
+
+WRITE_LINE_MEMBER(eispc_keyboard_device::rst_line_w)
+{
+	if (state == CLEAR_LINE)
+	{
+		m_mcu->resume(SUSPEND_REASON_RESET);
```

Comment:
Why not use `set_input_line` on the MCU to assert/clear its input line rather than abusing the scheduler?

Reply:
I don't see the reset input line being supported for the 6801/6803 currently? 

The RST line in the keyboard connector is hardwired to the RESET input at the 6801 so I figured it was important to actually not start the keyboard before the system was started as the keyboard sends a null byte on init. It looks like it is not critical though.

Reply:
See discussion below